mock/docs/mock-parse-buildlog.1
.TH "mock-parse-buildlog" "1" "@VERSION@" "Silvie Chlupova"
.SH NAME
mock-parse-buildlog \- parse build.log
.SH SYNOPSIS
.B mock-parse-buildlog
\fB\-\-path\fR \fIPATH\fR
.SH DESCRIPTION
.B mock-parse-buildlog
parses a build.log file saved in logs directory and tries to recognize what caused that a build failed and prints to stdout the cause.
.SH OPTIONS
.TP
\fB\-p\fR PATH, \fB\-\-path\fR \fIPATH\fR
set the path where is build.log located.

.SH "AUTHORS"
Silvie Chlupova


ChangeLog
2007-12-01  Michael Brown <mebrown@michaels-house.net>
	* configure.ac
	remove unnecessary configure items to speed configure
	* daemontest.c
	fix compile warnings
	* doc/releasetests.sh
	more comprehensive release tests
	* decorator.py
	remove our decorator.py in favor of std python-decoratortools
	* various
	pylint/pychecker recommendations
	make decorator syntax compatible with python 2.3 (for RHEL4 porting...)
	fix 'mock install'
	* mock.py
	popagate chroot/shell command exit codes to caller
	init groups list if we are called by sudo so we have access to
		all directories caller does
	

2007-12-01  Michael Brown <mebrown@michaels-house.net>
	* backend.py
	fixed build of multiple srpms on one cmdline

2007-11-29  Clark Williams  <williams@redhat.com>

	* mock.py, Makefile* 
	added back verbosity control options (-q and -v)
	added --offline option
	added dependency on python-cytypes
	lots of logging changes

	
2007-10-20  Michael Brown <mebrown@michaels-house.net>

	* mock.py
	lots of work redoing stuff. Now use setuid wrapper instead of
	setuid helper.
	huge number of changes upstream
	convert to setuid wrapper instead of old setuid helper
	lots of bugfixes and improvements
	/var/cache/yum now saved and bind-mounted
	ccache integration
	rootcache improvements (formerly called autocache)

2007-02-24  Clark Williams  <williams@redhat.com>

	* mock.spec
	BZ#226673 - Ville Skyttä fix for RPM_OPT_FLAGS

2007-02-20  Clark Williams  <williams@redhat.com>

	* mock.py
	Handle --no-clean option when doing yum.conf symlink
	
	* mock.spec 
	Bumped version to 0.6.13

2007-02-16  Clark Williams  <williams@redhat.com>

	* mock.py
	Added safety symlink to handle possible yum breakage in rawhide

2007-02-07  Clark Williams  <williams@redhat.com>

	* mock.py
	Added error() calls to print command output on failed commands

2007-02-06  Clark Williams  <williams@redhat.com>

	* mock.py
	added installdeps command for long-term chroot management
	fixed invalid reference to os.file (rather than os.path)
	
2007-01-08  Clark Williams  <williams@redhat.com>
	* Makefile
	changed archive creation to keep tarball clean

	* etc/fedora-4-*
	added Josh Boyer's EPEL config files

	* mock.spec
	bumped version to 0.6.10
	
2006-10-30  Clark Williams  <williams@redhat.com>

	* Makefile
	Bumped version to 0.6.7 for FC6 release

	* etc/fedora-6-*
	Added for FC6 release

2006-10-21  Clark Williams  <williams@redhat.com>

	* mock.spec
	Bumped version to 0.6.6

2006-09-11 11:21  jcwillia
	* mock.py
	added patch from Karanbir Singh to fix RPM problem in 
	CentOS 4.4

	* mock.spec
	bumped version to 0.6.5

2006-08-29 05:16  jcwillia

	* mock.py
	explicit import of rpmUtils.transaction 
	(BZ#204051)

2006-08-25 16:40  jcwillia

	* buildsys-build.spec
	Dennis Gilmore's patches for Aurora SPARC Linux 
	(BZ#204100)
	
2006-08-17 20:28  jcwillia
	
	* etc/redhat-73-i386.cfg, etc/redhat-9-i386.cfg:
	added runuser config option for legacy config files
	
	* mock.py
	updated usage message
	
	* doc/mock.1
	reworked man page to include caching options, chroot and shell
	commands. 
	(BZ#200080)

2006-08-17 20:23  jcwillia
	
	* etc/redhat-8-i386.cfg:
	added template for RHL8
	
2006-08-16 16:59  jcwillia

	* mock.spec
	added disttags
	
	* mock.py
	added check for basedir change and raise error when it's detected
	(BZ#1644441)
	
	
2006-08-07 17:26  jcwillia
	
	* buildsys-build.spec
	added rhel tags

	* Makefile
	Modified to build all supported targets for buildsys-build

2006-08-07 15:05  jcwillia

	* buildsys-build.spec
	Thorsten Leemhuis and Paul Howarth's patches for RHL and Legacy
	chroots.
	(BZ#196930)
	
2006-06-13 01:31  skvidal

	* etc/defaults.cfg: 
	add cache defaults

2006-06-13 01:16  skvidal

	* ChangeLog: 
	check in changelog

2006-06-13 01:15  skvidal

	* etc/defaults.cfg: 
	correct commented-out groupinstall option

2006-06-13 00:28  skvidal

	* etc/: fedora-3-i386-core.cfg, fedora-3-x86_64-core.cfg,
	fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-x86_64-core.cfg,
	fedora-devel-i386-core.cfg, fedora-devel-ppc-core.cfg,
	fedora-devel-x86_64-core.cfg: 
	config file changes to point to mirrors

2006-06-13 00:12  skvidal

	* etc/: fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-i386-core.cfg,
	fedora-5-ppc-core.cfg, fedora-5-x86_64-core.cfg: 
	fix up some of the url/mirrorlist paths

2006-06-12 23:52  skvidal

	* buildsys-build.spec: 
	reduce build set

2006-06-08 17:28  mebrown

	* mock.py, src/mock-helper.c: Make rm less verbose.

2006-06-07 08:36  skvidal

	* etc/: fedora-1-i386-core.cfg, fedora-2-i386-core.cfg,
	fedora-3-i386-core.cfg, fedora-3-x86_64-core.cfg,
	fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-i386-core.cfg,
	fedora-5-ppc-core.cfg, fedora-5-x86_64-core.cfg,
	fedora-devel-i386-core.cfg, fedora-devel-ppc-core.cfg,
	fedora-devel-x86_64-core.cfg, redhat-73-i386.cfg,
	redhat-9-i386.cfg: 
	fix up all the paths to buildgroups and to plague build-results

2006-06-07 08:29  skvidal

	* mock.py, mock.spec: 
	
	iterate version in spec files

2006-06-07 08:27  skvidal

	* etc/: fedora-1-i386-core.cfg, fedora-2-i386-core.cfg,
	fedora-3-i386-core.cfg, fedora-3-x86_64-core.cfg,
	fedora-4-i386-core.cfg, fedora-4-ppc-core.cfg,
	fedora-4-x86_64-core.cfg, fedora-5-i386-core.cfg,
	fedora-5-ppc-core.cfg, fedora-5-x86_64-core.cfg,
	fedora-devel-i386-core.cfg, fedora-devel-ppc-core.cfg,
	fedora-devel-x86_64-core.cfg, redhat-73-i386.cfg,
	redhat-9-i386.cfg: 
	clean out things that are now in 'defaults.cfg'

2006-06-06 17:28  skvidal

	* etc/defaults.cfg: use defaults more intelligently


2006-05-24 11:15  skvidal

	* mock.py, src/mock-helper.c: 
	check in Michael Brown's lovely autocache creation patch.

2006-05-17 11:12  jcwillia

	* mock.py: Michael E. Brown's patch for getting PS1 down to chroot
	shell

2006-05-16 00:15  skvidal

	* mock.py: applied tab-fix and a slimming of main patches from
	Michael E. Brown.


mock/py/mock-parse-buildlog.py
#!/usr/bin/python3 -tt
# -*- coding: utf-8 -*-
# vim:expandtab:autoindent:tabstop=4:shiftwidth=4:filetype=python:textwidth=0:

import argparse
import logging
import re

FORMAT = "%(levelname)s: %(message)s"
logging.basicConfig(format=FORMAT, level=logging.WARNING)
log = logging.getLogger()


def argumentParser():
    parser = argparse.ArgumentParser(
        description="Parses the build.log and return an error why build failed.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument("-p", "--path", required=True, help="Path to build.log")
    arguments = parser.parse_args()
    return arguments


def parseBuildLog(log_path):
    """Parses the build log.

    Args:
        log_path (str): Path to the RPM build log.

    Returns:
        tuple: The first element is the type of error that was found
            in the log (missing or deleted). The second element is
            a list of problematic files.
    """
    try:
        with open(log_path, "r") as build_log:
            lines = build_log.read().splitlines()
    except IOError as error:
        log.error("There was an error opening %s, %s", log_path, str(error))
        return

    error_re = re.compile(
        r"""
        ^
        (BUILDSTDERR:)?
        \s*
        (
            (?P<missing>File\s+not\s+found:\s*)|
            (?P<unpackaged>Installed\s+\(but\s+unpackaged\)\s+file\(s\)\s+found:)
        )?
        (?P<path>/.*)?
        $
        """, re.VERBOSE,
    )

    error_type = None
    files = set()

    for line in lines:
        match = error_re.match(line)
        if match:
            if match.group("missing"):
                error_type = "deleted"
                files.add(match.group("path"))
            elif match.group("unpackaged"):
                error_type = "missing"
            elif error_type == "missing" and match.group("path"):
                files.add(match.group("path"))
            elif error_type and not match.group("path"):
                break

    return error_type, list(files)


def main(log_path):
    error = parseBuildLog(log_path)
    if error[0] is not None:
        if error[0] == "missing":
            print(
                "Error type: {0}".format("Build failed because problematic files are in %buildroot but not in %files")
            )
        elif error[0] == "deleted":
            print(
                "Error type: {0}".format("Build failed because problematic files are in %files but not in %buildroot")
            )
        print("Problematic files: ")
        for files in error[1]:
            print(files)
    else:
        log.error("Couldn't recognize the error that caused the build failure.")


if __name__ == "__main__":
    programArguments = argumentParser()
    main(programArguments.path)
